### ISSUE TITLE
How to NOT run certain asp.net api middleware for proxied requests?

### ISSUE THREAD

**Issue by christianacca** [microsoft/reverse-proxy#1482]
#### TL;DR

I want to control which asp.net middleware applies to ONLY those requests directly serviced by say a `Controller` in an asp.net 6 api. Is there any guidance/feature for this in YARP?

#### Detail

I have an asp.net 6 api that is acting as a backend for frontend. This api will directly service some of the client requests itself, and some it will forward on (proxy) to other api's.

There will be a bunch of "regular" middleware that will be enriching the asp.net api middleware pipeline. It's highly likely that some of those middleware will need to be applied ONLY to the requests that the api will service directly ie these middleware should NOT run when the request is being proxied by YARP.

One way I can think of is with `UseWhen` EG: 

```c#
void ConfigureMiddleware(IApplicationBuilder app, IWebHostEnvironment environment) {
    if (environment.IsDevelopment()) {
        app.UseSwagger();
        app.UseSwaggerUI();
    }

    app.UseAuthentication();
    app.UseAuthorization();
    
    app.UseWhen(ctx => IsProxiedRequest(ctx) == false, OwnMiddleware);
}
```

However, I don't want to duplicate the rules for deciding if the request is to be proxied in both `IsProxiedRequest` method and the YARP routing rules.

### Why is this important to you?

I want to be in control of what code actually runs for each request hitting the backend api.

**Comment by samsp-msft** [comment]
I imagine it's going to depend on the ordering of the middleware and where the custom steps are positioned relative to UseRouting(). @Tratcher - Is there a way to know if a request has matched a YARP route before the YARP middleware kicks in?

**Comment by Tratcher** [comment]
Try this:
```
        private bool IsProxiedRequest(HttpContext context)
        {
            return context.GetEndpoint()?.Metadata.GetMetadata<RouteModel>() != null;
        }
```

See here for an example:
https://github.com/microsoft/reverse-proxy/blob/23d235fe9fde53419e8a40c2db02c8bdae972eaf/src/ReverseProxy/Model/ProxyPipelineInitializerMiddleware.cs#L28-L34

**Comment by christianacca** [comment]
Excellent, just what I was looking for :-)
